pkg/obj/merge.go

```diff
@@ -11,42 +11,49 @@ import (
 	"sigs.k8s.io/kustomize/kyaml/yaml/merge2"
 )
 
-type Merger interface {
-	Merge(src, dst runtime.Object) error
-}
-
-func NewMerger(scheme *runtime.Scheme) Merger {
+func NewSerializer(scheme *runtime.Scheme) runtime.Serializer {
 	codecs := serializer.NewCodecFactory(scheme)
-
 	info, _ := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), runtime.ContentTypeJSON)
-
-	return &merger{
-		serializer: info.Serializer,
-	}
+	return info.Serializer
 }
 
-type merger struct {
-	serializer runtime.Serializer
-}
+// Merge applies 'src' as a patch to 'dst', returning a new object containing the merge result.
+func Merge[T runtime.Object](source runtime.Object, dest T, serializer runtime.Serializer) (T, error) {
+	var empty T
 
-func (m *merger) Merge(src, dst runtime.Object) error {
 	var srcB bytes.Buffer
-	if err := m.serializer.Encode(src, &srcB); err != nil {
-		return fmt.Errorf("could not encode source obj: %w", err)
+	if err := serializer.Encode(source, &srcB); err != nil {
+		return empty, fmt.Errorf("could not encode source obj: %w", err)
 	}
 
 	var dstB bytes.Buffer
-	if err := m.serializer.Encode(dst, &dstB); err != nil {
-		return fmt.Errorf("could not encode destination obj: %w", err)
+	if err := serializer.Encode(dest, &dstB); err != nil {
+		return empty, fmt.Errorf("could not encode destination obj: %w", err)
 	}
 
-	out, err := strategicpatch.StrategicMergePatch(dstB.Bytes(), srcB.Bytes(), dst)
+	out, err := strategicpatch.StrategicMergePatch(dstB.Bytes(), srcB.Bytes(), dest)
 	if err == nil {
-		if _, _, err := m.serializer.Decode(out, nil, dst); err != nil {
-			return err
+		// Previously, we discarded the output and used 'dst' as 'into'. This introduced a bug in the following scenario:
```

Comment:
This is more of a PR comment than a code comment (referring to previous implementations). Can we move this to the PR comment and add a comment here about this case generally being handled?

Reply:
Perhaps even move it to the `Merge` godoc?